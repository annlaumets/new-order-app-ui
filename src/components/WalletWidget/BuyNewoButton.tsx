import { useState } from 'react'
import {
  AspectRatio,
  HStack,
  Link,
  MenuItem,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  Stack,
  Text,
  useDisclosure,
} from '@chakra-ui/react'
import Image from 'next/image'

import ModalOverlay from 'components/ModalOverlay'

import { exchanges } from 'constants/newo'

const BuyNewoButton = () => {
  const { isOpen, onOpen, onClose } = useDisclosure()
  const [overlay, setOverlay] = useState(<ModalOverlay />)

  return (
    <>
      <Modal isCentered isOpen={isOpen} onClose={onClose}>
        {overlay}
        <ModalContent>
          <ModalHeader fontSize="lg">Buy $NEWO</ModalHeader>
          <ModalCloseButton />

          <ModalBody>
            <Stack>
              {exchanges.map((exchange) => {
                return (
                  <Link
                    key={exchange.label}
                    href={exchange.link}
                    variant="orangeTransition"
                    color="white"
                    isExternal
                    fontSize="lg"
                    borderRadius="md"
                    p={['2', '4']}
                    _hover={{ bgColor: 'gray.80' }}
                  >
                    <HStack gap="1">
                      <AspectRatio ratio={1 / 1} pos="relative" w="10">
                        <Image
                          src={exchange.icon}
                          alt={exchange.label}
                          layout="fill"
                        />
                      </AspectRatio>

                      <Text>{exchange.label}</Text>
                    </HStack>
                  </Link>
                )
              })}
            </Stack>
          </ModalBody>
        </ModalContent>
      </Modal>

      <MenuItem
        onClick={() => {
          onOpen()
          setOverlay(<ModalOverlay />)
        }}
        _hover={{
          color: 'brand.orange',
        }}
      >
        Buy $NEWO
      </MenuItem>
    </>
  )
}

export default BuyNewoButton
