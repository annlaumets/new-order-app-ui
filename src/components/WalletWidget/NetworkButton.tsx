import {
  AspectRatio,
  Box,
  Divider,
  Flex,
  HStack,
  Image,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  Skeleton,
  Text,
  useToast,
} from '@chakra-ui/react'
import { useEffect, useState } from 'react'
import { useNetwork } from 'wagmi'

import ModalOverlay from 'components/ModalOverlay'

import { getIconForNetwork } from 'lib/utils/wallet'
import { env } from 'lib/environment'

import { useContractContext } from 'store/contexts/contractContext'

import { SUPPORTED_NETWORKS } from 'constants/network'

const NetworkButton = () => {
  const toast = useToast()
  const { isOpen, onOpen, onClose, hasNetworkError } = useContractContext()
  const [overlay, setOverlay] = useState(<ModalOverlay />)

  const {
    activeChain: networkData,
    error: switchNetworkError,
    isLoading,
    switchNetwork,
  } = useNetwork()

  const changeNetworkAndClose = (networkId: number) => {
    if (switchNetwork) {
      switchNetwork(networkId)
    }

    onClose()
  }

  useEffect(() => {
    if (switchNetworkError?.message) {
      toast({
        title: 'Switching Network Failed',
        description: switchNetworkError?.message,
        isClosable: true,
        position: 'top-right',
        status: 'error',
        variant: 'error',
      })
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [switchNetworkError?.message])

  return (
    <>
      <Flex
        onClick={() => {
          setOverlay(<ModalOverlay />)
          onOpen()
        }}
        alignItems="center"
        gap="3"
        ml={['2', 0]}
        flexWrap="wrap"
      >
        <Box
          w="2.5rem"
          h="2.5rem"
          border="1px"
          borderColor="gray.65"
          borderRadius="md"
          cursor="pointer"
          p="6px"
          transition="all 0.3s ease"
          _hover={{
            bgColor: 'gray.85',
          }}
          mr="2"
        >
          <AspectRatio ratio={1 / 1} position="relative">
            <Image
              borderRadius="md"
              bg="gray.85"
              src={getIconForNetwork(networkData?.id, 'source')}
              alt={getIconForNetwork(networkData?.id, 'alt')}
            />
          </AspectRatio>
        </Box>
        <Text display={['inline-flex', 'none']} fontSize="0.9rem">
          {networkData?.name}
        </Text>
      </Flex>
      <Modal
        isCentered
        motionPreset="scale"
        isOpen={hasNetworkError ? hasNetworkError : isOpen}
        onClose={onClose}
        closeOnOverlayClick={hasNetworkError ? hasNetworkError : true}
      >
        {overlay}
        <ModalContent
          bgColor="gray.100"
          color="white"
          border="1px"
          borderRadius="md"
          borderColor="gray.90"
        >
          <ModalHeader fontSize="0.9rem">
            Select a supported network
          </ModalHeader>
          <ModalCloseButton
            mt="0.5"
            _focus={{ boxShadow: 'none' }}
            _hover={{ color: 'brand.red' }}
          />
          <Divider />
          <ModalBody py="4" w="full">
            {isLoading
              ? Array.from(Array(3).keys()).map((i: number) => {
                  return (
                    <Skeleton
                      key={i}
                      borderRadius="sm"
                      h="50px"
                      w="full"
                      startColor="gray.80"
                      endColor="gray.85"
                      mb="2"
                    />
                  )
                })
              : switchNetwork &&
                SUPPORTED_NETWORKS.filter((x) => {
                  if (env.isDev || env.isDevelopment || env.isTest) {
                    return x
                  } else if (env.isProd || env.isProduction) {
                    return x.testnet === false
                  }
                  return x.testnet === false
                }).map((x) => {
                  return x.id === networkData?.id ? null : (
                    <HStack
                      onClick={() => {
                        changeNetworkAndClose(x.id)
                      }}
                      key={x.id}
                      _hover={{
                        backgroundColor: 'gray.85',
                      }}
                      transition="0.3s ease all"
                      cursor="pointer"
                      p="4"
                      borderRadius="md"
                    >
                      <Box w="2rem" h="2rem" mr="2">
                        <AspectRatio ratio={1 / 1} position="relative">
                          <Image
                            borderRadius="md"
                            bg="gray.85"
                            src={getIconForNetwork(x.id, 'source')}
                            alt={getIconForNetwork(x.id, 'alt')}
                          />
                        </AspectRatio>
                      </Box>
                      <Text fontSize="0.9rem">
                        {'Switch to '}
                        {x.name}
                      </Text>
                    </HStack>
                  )
                })}
          </ModalBody>
        </ModalContent>
      </Modal>
    </>
  )
}

export default NetworkButton
