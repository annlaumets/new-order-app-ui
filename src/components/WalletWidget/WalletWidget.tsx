import {
  Box,
  Button,
  Divider,
  Icon,
  Text,
  useDisclosure,
} from '@chakra-ui/react'
import {
  Modal,
  ModalContent,
  ModalHeader,
  ModalBody,
  ModalCloseButton,
} from '@chakra-ui/react'
import { useState } from 'react'
import { useConnect } from 'wagmi'
import MetaMaskOnboarding from '@metamask/onboarding'

import AddressButton from 'components/WalletWidget/AddressButton'
import ModalOverlay from 'components/ModalOverlay'

import { getIconForConnector } from 'lib/utils/wallet'

import { useNewoContext } from 'store/contexts/newoContext'

const WalletWidget = ({ ...rest }) => {
  const { isOpen, onOpen, onClose } = useDisclosure()
  const { metamaskIsInstalled, accountAddress } = useNewoContext()

  const [overlay, setOverlay] = useState(<ModalOverlay />)
  const { connectors, isLoading, connectAsync } = useConnect()

  const getBrowserPlugin = () => {
    const onboarding = new MetaMaskOnboarding()

    onboarding.startOnboarding()
  }

  return (
    <>
      <Box {...rest}>
        {accountAddress ? (
          <AddressButton />
        ) : (
          <Button
            variant="greenButton"
            onClick={() => {
              setOverlay(<ModalOverlay />)
              onOpen()
            }}
            loadingText="Connecting Wallet"
            isLoading={isLoading}
          >
            Connect Wallet
          </Button>
        )}
      </Box>
      <Modal isCentered motionPreset="scale" isOpen={isOpen} onClose={onClose}>
        {overlay}
        <ModalContent
          bgColor="gray.100"
          color="white"
          border="1px"
          borderRadius="md"
          borderColor="gray.90"
        >
          <ModalHeader fontSize="0.9rem">Connect Wallet</ModalHeader>
          <ModalCloseButton
            mt="0.5"
            _focus={{ boxShadow: 'none' }}
            _hover={{ color: 'brand.red' }}
          />
          <Divider />
          <ModalBody py="4">
            {connectors.map((connector) => {
              const connectorName =
                connector.id === 'injected'
                  ? connector.name
                    ? connector.name === 'Injected'
                      ? 'Metamask'
                      : connector.name
                    : connector.id
                  : connector.name === 'Injected'
                  ? 'Metamask'
                  : connector.name
              return (
                <Button
                  key={connector.id}
                  disabled={!connector.ready && connector.id !== 'injected'}
                  leftIcon={
                    <Icon
                      mr="4"
                      fontSize="36"
                      fill="white"
                      as={getIconForConnector(connector.id)}
                    />
                  }
                  variant="walletButton"
                  onClick={() => {
                    if (!connector.ready && !metamaskIsInstalled) {
                      getBrowserPlugin()
                    } else {
                      connectAsync({ connector })
                    }
                    onClose()
                  }}
                >
                  {connector.id === 'injected' && !metamaskIsInstalled ? (
                    <Text fontSize="sm">Download Metamask extension</Text>
                  ) : (
                    <>
                      <Text
                        fontSize={
                          connector.name?.includes('Injected') ? 'xs' : 'lg'
                        }
                      >
                        {connectorName}
                      </Text>
                      {!connector.ready && ' (unsupported)'}
                      {isLoading &&
                        connector.name === connector?.name &&
                        '(connecting)'}
                    </>
                  )}
                </Button>
              )
            })}
          </ModalBody>
        </ModalContent>
      </Modal>
    </>
  )
}

export default WalletWidget
