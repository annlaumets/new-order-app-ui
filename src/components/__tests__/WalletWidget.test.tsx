import { render } from '@testing-library/react'
import { WagmiConfig } from 'wagmi'
import { ChakraProvider } from '@chakra-ui/react'
import { defaultChains, MockConnector, wallets } from 'wagmi-testing'

import WalletWidget from 'components/WalletWidget/WalletWidget'

import theme from 'theme'

describe('WalletWidget.tsx component', () => {
  const connector: MockConnector = new MockConnector({
    chains: defaultChains,
    options: {
      network: 1,
      privateKey: wallets.ethers1.privateKey,
    },
  })

  test('match to snapshot', () => {
    const { container } = render(
      <ChakraProvider theme={theme}>
        {/*  eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
        {/* @ts-ignore */}
        <WagmiConfig connectors={[connector]}>
          <WalletWidget />
        </WagmiConfig>
      </ChakraProvider>
    )

    expect(container).toMatchSnapshot()
  })

  test('match to snapshot with connected account', () => {
    const { container } = render(
      <ChakraProvider theme={theme}>
        {/*  eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
        {/* @ts-ignore */}
        <WagmiConfig connectors={[connector]} autoConnect>
          <WalletWidget />
        </WagmiConfig>
      </ChakraProvider>
    )

    expect(container).toMatchSnapshot()
  })
})
