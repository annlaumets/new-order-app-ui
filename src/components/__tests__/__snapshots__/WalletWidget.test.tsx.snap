// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletWidget.tsx component match to snapshot 1`] = `
<div>
  <div
    class="css-0"
  >
    <button
      class="chakra-button css-1iooehf"
      type="button"
    >
      Connect Wallet
    </button>
  </div>
  <span
    class="chakra-env"
    hidden=""
  />
</div>
`;

exports[`WalletWidget.tsx component match to snapshot with connected account 1`] = `
<div>
  <div
    class="css-0"
  >
    <button
      class="chakra-button css-1iooehf"
      type="button"
    >
      Connect Wallet
    </button>
  </div>
  <span
    class="chakra-env"
    hidden=""
  />
</div>
`;
